src/lambda/utilities/getMainSourceFile.ts

```diff
@@ -0,0 +1,120 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as os from 'os'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { CloudFormation } from '../../shared/cloudformation/cloudformation'
+import * as fs from '../../shared/filesystemUtilities'
+import { filterAsync, first } from '../../shared/utilities/collectionUtils'
+import { detectLocalTemplates } from '../local/detectLocalTemplates'
+import { isNodeJS } from '../models/samLambdaRuntime'
+
+export interface OpenMainSourceFileUriContext {
+    getLocalTemplates(...workspaceUris: vscode.Uri[]): AsyncIterable<vscode.Uri>
+    loadSamTemplate(uri: vscode.Uri): Promise<CloudFormation.Template>
+    fileExists(path: string): Promise<boolean>
+}
+
+export async function getMainSourceFileUri(
+    {
+        root,
+        getLocalTemplates = (...workspaceUris: vscode.Uri[]) => detectLocalTemplates({ workspaceUris }),
+        loadSamTemplate = async uri => await CloudFormation.load(uri.fsPath),
+        fileExists = fs.fileExists
+    }: Partial<OpenMainSourceFileUriContext> & {
+        root: vscode.Uri
+    }
+): Promise<vscode.Uri> {
+    const templateUri = await first(getLocalTemplates(root))
+    if (!templateUri) {
+        throw new Error(`Invalid project format: '${root.fsPath}' does not contain a SAM template.`)
+    }
+
+    const lambdaResource: CloudFormation.Resource = await getFirstLambdaResource({
+        templateUri,
+        loadSamTemplate,
+        fileExists
+    })
+
+    return await getSourceFileUri({
+        root: vscode.Uri.file(path.dirname(templateUri.fsPath)),
+        resource: lambdaResource,
+        fileExists
+    })
+}
+
+async function getFirstLambdaResource(
+    {
+        templateUri,
+        loadSamTemplate
+    }: Pick<OpenMainSourceFileUriContext, 'loadSamTemplate' | 'fileExists'> & {
+        templateUri: vscode.Uri
+    }
+): Promise<CloudFormation.Resource> {
+    const template = await loadSamTemplate(templateUri)
+    if (!template.Resources) {
+        throw new Error(`SAM Template '${templateUri.fsPath}' does not contain any resources`)
+    }
+
+    const lambdaResources = Object.getOwnPropertyNames(template.Resources)
+        .map(property => template.Resources![property])
+        .filter(resource => resource.Type === 'AWS::Serverless::Function')
```

Comment:
Would it also make sense to filter where `resource.Properties` exists?

Reply:
Possibly. But I think that `SAM Template '${templateUri.fsPath}' does not contain any lambda resources`/`Lambda resource is missing the 'Properties' property:${os.EOL}` are better error messages than `SAM Template '${templateUri.fsPath}' does not contain any lambda resources with properties`

Aside from the message, the only way that it would matter is if SAM init's default template changed to include a lambda resource (with no properties), then another lambda resource (with properties). This would be weird and probably unintentional, so I think it's better to throw.